lib/queuePopulator/RaftLogReader.js

```diff
@@ -19,8 +28,210 @@ class RaftLogReader extends LogReader {
         super({ zkClient, kafkaConfig, logConsumer, logId: `raft_${raftId}`,
                 logger, extensions, metricsProducer });
         this.raftId = raftId;
+        this._iProducer = new IngestionProducer(bucketdConfig);
+        this.newIngestion = false;
+        this.removeLogOffset = null;
+    }
+
+    /* eslint-disable no-param-reassign */
+
+    _processReadRecords(params, batchState, done) {
```

Comment:
It seems to me the core logic should be reused from `LogReader` code, not copied over and modified.

Something that may work better to do this is adding some specific control points in LogReader where the Raft backend can be notified and do its own bookkeeping tasks (with optional methods or event handlers), without interfering with the core logic of the log reader. It should be more maintainable than having entire private core functions copied in the subclass for specific handling.

I understand it may not be that easy to get the proper code architecture but on the long run it pays off.